### PR TITLE
feat(plugin-adblocker): Add cooperative intercept mode support

### LINKED ISSUE #364: Solving request interception chain
Hi Everyone!

Currently I am having the problem that when you have 2 plugins that try to intercept and handle requests, that the dependencies start to fail. (E.g. when I have the adblocker plugin and a personal plugin that both try to block resources, then it starts stating that a request has already been handled).

I agree that this is how it is with Puppeteer and that it will be hard to fix it (sadly enough). But seeing the plugin interface, I think it might be worthwhile to find a solution to this.

Some thoughts that I had:
* Create page hooks (such as the onPageCreated) but then for onPageRequest, onPageRequestFinished, onPageRequestFailed and make them check if a request has been handled already (it should be possible to keep working with requests until an .abort has been called or the request finished?)
* Add a weight system to allow certain plugins to come before others (e.g. adblocker to block requests before other plugins).

I'd love to hear some thoughts from the community to see if this is feasible or not? Looking forward to the responses!

### PR DESCRIPTION
**Pull request by FdezRomero**
Adds support for the new [Cooperative Intercept Mode available in Puppeteer v11+](https://github.com/puppeteer/puppeteer/blob/v11.0.0/docs/api.md#upgrading-to-cooperative-mode-for-package-maintainers) and defining a custom `interceptResolutionPriority` for the following plugins:
- `puppeteer-extra-plugin-block-resources`
- `puppeteer-extra-plugin-block-adblocker`

This change checks that the required interception functions are available, so it's backwards compatible with the Legacy Mode and previous Puppeteer versions.

The `@cliqz/adblocker-puppeteer` dependency [already included support for Cooperative Mode in v1.23.0](https://github.com/cliqz-oss/adblocker/blob/master/packages/adblocker-puppeteer/CHANGELOG.md#v1230-mon-oct-18-2021).

Fixes #364, #156

### CODE REVIEW

File: packages/puppeteer-extra-plugin-adblocker/src/index.ts
Comment (benallfree) [inline packages/puppeteer-extra-plugin-adblocker/src/index.ts:44]: I recommend defaulting to `undefined` to preserve legacy mode behavior. A priority should be explicitly set by the user if they want to use cooperative mode - at least, for now. Otherwise users may experience the handlers resolving in a different order than before, and that can be confusing.

tldr; Default to legacy mode and add a note to the README about how to enable cooperative mode for users who want it.

File: packages/puppeteer-extra-plugin-block-resources/index.js
Comment (benallfree) [inline packages/puppeteer-extra-plugin-block-resources/index.js:69]: Same note: `undefined` default

**Review by benallfree — CHANGES_REQUESTED** [review]

File: packages/puppeteer-extra-plugin-block-resources/index.js
Comment (benallfree) [inline packages/puppeteer-extra-plugin-block-resources/index.js:136]: For backward compatibility, consider `request.continueRequestOverrides ? request:continueRequestOverrides() : undefined`. Same idea with `.abort()` and `.respond()`

**Review by benallfree — CHANGES_REQUESTED** [review]

File: packages/puppeteer-extra-plugin-adblocker/src/index.ts
Comment (FdezRomero) [reply packages/puppeteer-extra-plugin-adblocker/src/index.ts:44]: Good point! I'll change it to be disabled by default 👍🏻

**Comment by FdezRomero** [comment]
@benallfree I implemented your suggestions and fixes. I also included a TypeScript definition that I was using on my client app 😉 Let me know if you find something else.

File: packages/puppeteer-extra-plugin-block-resources/index.js
Comment (benallfree) [inline packages/puppeteer-extra-plugin-block-resources/index.js:133]: I think I missed this the first time. Can you walk me through the intent? In cooperative mode, the Puppeteer core will run all handlers and resolve the intercept to the one with the highest priority. Unless you have a special case, your handler should simply do its computation and respond at whatever priority it is set to. If a different response at a higher priority takes place, it will prevail and yours will “lose”. If yours is the highest priority response, it will “win”

**Review by benallfree — CHANGES_REQUESTED** [review]

File: packages/puppeteer-extra-plugin-block-resources/index.js
Comment (FdezRomero) [reply packages/puppeteer-extra-plugin-block-resources/index.js:133]: @benallfree The problem is that if you use a plugin that hasn't been migrated to Cooperative Mode and still uses Legacy Mode, the request is handled right away.

So if we call to `continue`/`abort`/`respond` etc. then we get the `Request is already handled!` error again (I just tried it again to verify). That's why we need this condition here, because we can't asume all plugins are opting-in to the new mode.

Comment (benallfree) [reply packages/puppeteer-extra-plugin-block-resources/index.js:133]: Oh that’s very smart, I think I’ll suggest this pattern in the official docs! Well done.

**Review by benallfree — APPROVED** [review]

**Comment by FdezRomero** [comment]
@benallfree Glad to hear that! It seems that this PR needs a second approval. Do you know who can help with that? Thanks!

**Comment by benallfree** [comment]
@FdezRomero I'm not a maintainer of this package so I'm not sure how to help with a 2nd approval.

By the way, I just submitted https://github.com/puppeteer/puppeteer/pull/7795/files based on your work here. It fixes that spelling mistake and officially exposes `isInterceptResolutionHandled()` logic. You might consider waiting on this merge until we know whether the Puppeteer team will accept my PR.

**Comment by benallfree** [comment]
@FdezRomero Small note, cooperative support came in 10.2 https://github.com/puppeteer/puppeteer/blob/v11.0.0/CHANGELOG.md#1020-2021-08-04

**Comment by benallfree** [comment]
@FdezRomero What will break?  #7796 does not introduce breaking changes, or it’s not supposed to :)

**Comment by FdezRomero** [comment]
You're right @benallfree, `@cliqz/adblocker-puppeteer` isn't using the `interceptResolution()` method so it won't be affected by the rename to `interceptResolutionState()`. It doesn't use it to check if the request has been already handled or not 👍🏻

I moved this PR back to draft and will apply the changes once https://github.com/puppeteer/puppeteer/pull/7796 ships. Hopefully it gets reviewed soon!.

**Comment by benallfree** [comment]
@FdezRomero  puppeteer/puppeteer#7796 has been accepted by the Puppeteer team 👍

**Comment by FdezRomero** [comment]
@benallfree Great! I'll update the PR with the new API 👍🏻
Any idea when a new version might be released?

**Comment by benallfree** [comment]
@FdezRomero No idea 😅

**Comment by FdezRomero** [comment]
Hi @benallfree, I have updated `adblocker-puppeteer` to the latest version with the fixes and upgraded the code to your newer Puppeteer API, which also made the code easier to read. Would you mind giving it another look? Thanks! 🙏🏻

File: packages/puppeteer-extra-plugin-block-resources/index.js
Comment (benallfree) [inline packages/puppeteer-extra-plugin-block-resources/index.js:129]: Are you using `continueRequestOverrides` purely for version detection? If so, it might make more sense to use  [abortErrorReason](https://github.com/puppeteer/puppeteer/blob/dc23b7535cb958c00d1eecfe85b4ee26e52e2e39/src/common/HTTPRequest.ts#L220) since this is an `abort()` call.

File: packages/puppeteer-extra-plugin-adblocker/src/index.ts
Comment (benallfree) [inline packages/puppeteer-extra-plugin-adblocker/src/index.ts:119]: Not familiar enough with the lib to know if this is correct, it seems to be passing it down a layer?

**Review by benallfree — APPROVED** [review]
lgtm w optional suggestions 👍

File: packages/puppeteer-extra-plugin-adblocker/src/index.ts
Comment (FdezRomero) [reply packages/puppeteer-extra-plugin-adblocker/src/index.ts:119]: @benallfree Yes exactly, it references the instance of `adblocker-puppeteer` that is used internally, which is the one that really intercepts and handles the requests, and this `pupeteer-extra` plugin acts as a wrapper: https://github.com/ghostery/adblocker/blob/master/packages/adblocker-puppeteer/adblocker.ts#L282-L283

File: packages/puppeteer-extra-plugin-adblocker/readme.md
Comment (FdezRomero) [reply packages/puppeteer-extra-plugin-adblocker/readme.md]: Yes, it's exported. I changed it in all examples 👍🏻

File: packages/puppeteer-extra-plugin-block-resources/index.js
Comment (FdezRomero) [reply packages/puppeteer-extra-plugin-block-resources/index.js]: I didn't know about this method, thanks! I changed it for the feature detection 👍🏻

**Comment by FdezRomero** [comment]
I think it's good to go! Let's see if it can get a bit of love from the maintainers 🤞🏻😄 @berstend

**Comment by FdezRomero** [comment]
Happy new year @berstend! 🎉

How can we get a second review to get this merged? We already made the necessary changes upstream in Puppeteer and Adblocker to support Cooperative Intercept mode in `puppeteer-extra` 😉

Thanks!

**Review by benallfree — APPROVED** [review]

**Review by deldrid1 — APPROVED** [review]

**Comment by deldrid1** [comment]
@berstend - Any chance this can be reviewed? This is a critical feature I need!

**Comment by benallfree** [comment]
@deldrid1 @berstend You both might consider running your own fork for a while, it seems the maintainer is busy on other things.

**Comment by berstend** [comment]
```
Successfully published:
 - puppeteer-extra-plugin-adblocker@2.13.0
 - puppeteer-extra-plugin-block-resources@2.4.0
 ```

**Pull request merged**
